include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,709 @@
+{
+  /*! @addtogroup Vectors
+   *@{
+   */
+
+  /**
+   * ReadWriteVector stores a subset of all elements. It allows to access
+   * individual elements to read or to write. However, it does not allow global
+   * operations such as taking the norm. ReadWriteVector can be used to read and
+   * write elements in vector derived from VectorSparceVector such as
+   * TrilinosWrappers::MPI::Vector and PETScWrappers::MPI::Vector.
+   *
+   * @author Bruno Turcksin, Wolfgang Bangerth, 2015.
+   */
+  template <typename Number>
+  class ReadWriteVector : public Subscriptor
+  {
+  public:
+    /**
+     * Declare standard types used in all containers. These types parallel
+     * those in the <tt>C++</tt> standard libraries <tt>vector<...></tt>
+     * class.
+     */
+    typedef Number                                            value_type;
+    typedef value_type                                       *pointer;
+    typedef const value_type                                 *const_pointer;
+    typedef value_type                                       *iterator;
+    typedef const value_type                                 *const_iterator;
+    typedef value_type                                       &reference;
+    typedef const value_type                                 &const_reference;
+    typedef types::global_dof_index                           size_type;
+    typedef typename numbers::NumberTraits<Number>::real_type real_type;
+
+    /**
+     * @name 1: Basic Object-handling
+     */
+    //@{
+    /**
+     * Empty constructor.
+     */
+    ReadWriteVector ();
+
+    /**
+     * Copy constructor.
+     */
+    ReadWriteVector (const ReadWriteVector<Number> &in_vector);
+
+    /**
+     * Constructs a vector given the size, the stored elements have their index
+     * in [0,size).
+     */
+    ReadWriteVector (const size_type size);
+
+    /**
+     * Constructs a vector whose stored elements indices are given by an
+     * IndexSet.
+     */
+    ReadWriteVector (const IndexSet &local_range);
+
+    /**
+     * Destructor.
+     */
+    ~ReadWriteVector ();
+
+    /**
+     * Sets the global size of the vector to @p size. The stored elements have
+     * their index in [0,size).
+     */
+    void reinit (const size_type size,
+                 const bool      fast = false);
+
+    /**
+     * Uses the same IndexSet as the one of the input vector @p in_vector and
+     * allocates memory for this vector.
+     *
+     * If the flag @p fast is set to false, the memory will be initialized
+     * with zero, otherwise the memory will be untouched (and the user must
+     * make sure to fill it with reasonable data before using it).
+     */
+    template <typename Number2>
+    void reinit(const ReadWriteVector<Number2> &in_vector,
+                const bool             fast = false);
+
+    /**
+     * Initializes the vector. The indices are specified by @p local_range,
+     */
+    void reinit (const IndexSet &local_range);
```

Comment:
Also, maybe not call the index set `local_range` but `locally_stored_indices` or similar. Same for the corresponding constructor above.